hibernate-core/src/test/java/org/hibernate/jpa/test/graphs/EntityGraphAttributeResolutionTest.java

```diff
@@ -0,0 +1,254 @@
+package org.hibernate.jpa.test.graphs;
+
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.persistence.CollectionTable;
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.EntityGraph;
+import javax.persistence.EntityManager;
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToMany;
+import javax.persistence.NamedAttributeNode;
+import javax.persistence.NamedEntityGraph;
+import javax.persistence.Table;
+
+import org.hibernate.graph.GraphSemantic;
+import org.hibernate.jpa.test.BaseEntityManagerFunctionalTestCase;
+
+import org.hibernate.testing.TestForIssue;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+
+/**
+ * @author Benjamin M.
+ * @author Nathan Xu
+ */
+@TestForIssue( jiraKey = "HHH-14113" )
+public class EntityGraphAttributeResolutionTest extends BaseEntityManagerFunctionalTestCase {
+
+	private User u;
+	private Group g1, g2;
+
+	@Override
+	protected Class<?>[] getAnnotatedClasses() {
+		return new Class[] { User.class, Group.class };
+	}
+
+	@Before
+	public void setUp() {
+		EntityManager em = getOrCreateEntityManager();
+		em.getTransaction().begin();
+
+		g1 = new Group();
+		g1.addPermission( Permission.BAR );
+		em.persist( g1 );
+
+		g2 = new Group();
+		g2.addPermission( Permission.BAZ );
+		em.persist( g2 );
+
+		u = new User();
+		em.persist( u );
+		u.addGroup( g1 );
+		u.addGroup( g2 );
+
+		em.getTransaction().commit();
+	}
+
+	@Test
+	public void fetchAssocWithNamedFetchGraph() {
+		EntityManager em = getOrCreateEntityManager();
+		List result = em.createQuery( "SELECT u.groups FROM User u WHERE u.id = ?1" )
+				.setParameter(1, u.getId() )
+				.setHint( GraphSemantic.FETCH.getJpaHintName(), em.getEntityGraph( Group.ENTITY_GRAPH ) )
+				.getResultList();
+
+		assertEquals( Arrays.asList( g1, g2 ), result );
```

Comment:
I noticed there seems to be some additional issue here. So from the HQL it seems the `result` should be a list of `Set<Group>`, but the real result is a list of `Group`. Is that a bug?

Reply:
You are mixing things up. When encountering a path expression that ends on a non-basic typed attribute, it is implicit joined. So the query would be equivalent to the following `SELECT g FROM User u JOIN u.groups g WHERE u.id = ?1`

Reply:
thanks for clarification. my knowledge gap here.